### ISSUE TITLE
Invalid session state: stored %!s(<nil>) on iOS Safari

### ISSUE THREAD

**Issue by dansjackson** [vouch/vouch-proxy#323]
I'm having an issue whereby I keep getting 400 Bad Request on iOS Safari - iOS Chrome works normally as do desktop browsers (most of the time)

I've had to disable testing mode because the issue does not happen on testing mode. Once I click through the links on testing mode, my application loads normally. 

Basically the error is:

`Invalid session state: stored %!s(<nil>)`

This happens after my OpenID provider (Keycloak) redirects the browser to the vouch.XXX.com/auth domain. As far as I can tell, because of this 400 Error, no JWT token is set by Vouch because if I try to manually browse to the protected domain, I get the 400 error again. The OpenID seems to authenticate fine however but Vouch doesn't accept the redirect from Keycloak

A little bit about my setup:

All sub domains are behind Cloudflare though the issue happened before Cloudflare so probably not relevant
All services are hosted on the same server.

Vouch is within a docker container with bridge networking port forwarded to 9090
Keycloak is within a docker container with bridge networking though auth requests go through nginx reverse proxy
Nginx is on the docker host but not in a container
Nginx proxies to my protected app which is also in a container.

Here are the details:
https://gist.github.com/dansjackson/fe54bc2a1e94d18101e10a30fa739224

**Desktop (please complete the following information):**
All desktop browsers work generally - Chrome sometimes has an issue but the JWT is set so if I refresh the page a couple of times it loads my protected apps eventually.

**Smartphone (please complete the following information):**
- iOS 14 - iPhone
- Safari doesn't work - Chrome on iOS works first time every time.

Things I've tried:
- cookie: secure: false (with samesite set to lax)
- commented out a load of extra stuff in nginx (See the above details "Vouch Reverse Nginx Conf")
- Tried to proxy pass directly to the domain name on the /validate (Got SSL error in nginx error log) - Validate seems to work best when nginx directly connects to the docker container rather than nginx trying to validate vouch.XXX.com which is also on nginx. 
- Enabled test mode: Fixes issue but I can't keep it on forever because its annoying. 
- Rather than setting $http_host, I've tried to specify the host manually too with no luck

This is how I understand the flow:

Safari visits protected app via Mobile data (outside the local network)
nginx on protected app auth requests on the /validate url
the /validate url proxy passes to vouch on localhost which returns 401
nginx gets the 401 and 301s Safari to https://vouch.xxx.com/login
After setting a session state on /login, /login 302's to keycloak
Safari logs into Keycloak
Keycloak 302s Safari to https://vouch.xxx.com/auth with the GET variables listed in the logs above
Vouch doesn't accept Safari's session anymore? Maybe Safari has changed something along the way? Maybe nginx is filtering something?

I understand this is not a lot to go off but because it's only in iOS on Safari, it's quite difficult to get any more information about the session storage and the cookies set on iOS. 

Let me know if I can provide any further info.

**Comment by bnfinet** [comment]
@dansjackson please provide logs as described in the README with `debug` and `testing`.

**Comment by dansjackson** [comment]
Hey @bnfinet,

Thanks for getting back to me. As requested, here are the complete docker logs from container startup to the 400 Error.

Interestingly when trying testing mode a second time, I am also now getting the 400 error on this too.

Debug:

https://gist.github.com/dansjackson/172b3f69f6632d8709a5697b81a81055

Debug with Testing Mode enabled:

https://gist.github.com/dansjackson/2fddd5320abe9572e2b225ccea684733

When I try to run: docker exec -it container do.sh bug_report - I get that the file is not found. I don't know where do.sh is within the container to provide the logs this way. 

For further context, I launched the container from:

docker run -d \
    -p 9090:9090 \
    --name vouch-proxy \
    -v ${PWD}/config:/config \
    voucher/vouch-proxy

So I don't think I've got the do.sh file on my server so I was not able to create the logs using this method and have instead taken the logs directly from "docker log"

Thanks,

**Comment by bnfinet** [comment]
@dansjackson those show `LogLevel:info`  Could you please try again with `debug`

and please post your yaml config

**Comment by dansjackson** [comment]
@bnfinet  Apologies, you are right. I didn't realise I had to set an additional docker env variable. Here you go!

Actual Debug Log:
https://gist.github.com/dansjackson/dce10cc795f15e91aa5abc2d2aab8300

Actual Debug Log with Testing:
https://gist.github.com/dansjackson/4e120db1ce48a6528f35866bc8090d74

I've copied the contents of the container out into a folder on my system using docker cp:

`docker cp vouch-proxy:/ /opt/docker-confs/vouch/`

I've searched the folder and cannot find a config.yml file for you. 

```
root@hostname:/opt/docker-confs/vouch# find . -name '*.yml'
./.defaults.yml
```

All of the config changes I have made are to docker environment variables which I have listed in the first gist.

Please let me know if I'm missing something

**Comment by dansjackson** [comment]
@bnfinet Thanks for getting back to me. It appears to be intermittent more than anything so I think the fact its Safari might be a red herring. I think that it worked on Safari in testing mode the first time however other tests in testing mode afterwards, it didn't work. The issue also sometimes happens on Chrome desktop, though again this intermittent and certainly way less frequent than Safari which seems to have the issue most of the time.

As requested however, I will try out your suggestions and get back to you. I don't know how far I'm going to get if I disable the Cookie domain however as this is needed to get me past even the validation stage but I'll give it a go and report back with some logs

**Comment by bnfinet** [comment]
@dansjackson also try..
* removing domain and using `VOUCH_ALLOWALLUSERS=true`

**Comment by ibotheperfect** [comment]
I have same issue. I'm using docker image 0.17.3 and chrome

**Comment by bnfinet** [comment]
@ibotheperfect please provide logs and config in the manner described in the README

@dansjackson any updates?

**Comment by dansjackson** [comment]
> @ibotheperfect please provide logs and config in the manner described in the README
> 
> @dansjackson any updates?

Literally just started testing about 20 mins ago. I'll provide an update shortly.

**Comment by dansjackson** [comment]
Ok something has fixed somewhere:

```
VOUCH_COOKIE_sameSite: lax
VOUCH_TESTING: false
VOUCH_ALLOWALLUSERS: true
VOUCH_DOMAINS: (DELETED)
```

Seems to be working. Let me try and narrow down which option it was

**Comment by dansjackson** [comment]
OK, this may highlight the intermittent nature of the problem. I've reapplied all of the settings that I removed and the issue appears to be fully resolved. I'm not able to replicate the error any more.

Could be that it re-ordered the variables in Docker (not sure if that makes a difference)
or
Could be just that the issue is not happening right now (though previously I was always able to replicate it on Safari)?
or
It could have just been my iPhone (Previously had an iPhone 8 Plus, now have iPhone 12 Pro).

Either way, the issue appears to be resolved on my end so from my perspective, the issue can be closed off. 

I don't think its the iPhone change because I was having the issue earlier today in Chrome for Windows, but I've deleted the variables and relaunched the container from the same image, then re-added the variables and relaunched again and something has fixed it. 

I know this won't help anyone else given how vague this answer is. I made no changes in the end to nginx as I didn't need to when I noticed the issue resolve. Here is a link to the order of the variables incase this was the fix:

https://gist.github.com/dansjackson/b3bcd286f1374274e531afa3e4bd01e1

I will keep an eye on it and will comment on the issue again if it re-occurs. I'm anticipating it might but hoping that it won't. Thanks for all the help @bnfinet
